### ISSUE TITLE
SIGSEGV when closing `ZstdCompressCtx`

### ISSUE THREAD

**Issue by gaojieliu** [luben/zstd-jni#331]
There is a race condition in `AutoCloseBase#close` based on my understanding
https://github.com/luben/zstd-jni/blob/master/src/main/java/com/github/luben/zstd/AutoCloseBase.java#L68
```
@Override
    public void close() {
        // Note: still should use synchronized in addition to sharedLock, because this class must support racy close(),
        // the second could happen through finalization or cleaning (when Cleaner is used). When updated to Java 9+,
        // synchronization block could be replaced with try { ... } finally { Reference.reachabilityFence(this); }
        synchronized (this) {
            if (sharedLock == SHARED_LOCK_CLOSED) {
                return;
            }
            if (!SHARED_LOCK_UPDATER.compareAndSet(this, 0, SHARED_LOCK_CLOSED)) {
                throw new IllegalStateException("Attempt to close while in use");
            }
            doClose();
        }
    }
```
Some other compressing thread can increase the `sharedLock` between
```
            if (!SHARED_LOCK_UPDATER.compareAndSet(this, 0, SHARED_LOCK_CLOSED)) {
                throw new IllegalStateException("Attempt to close while in use");
            }
```
and
```
            doClose();
```
Since `doClose` would deallocate the underlying native reference, the other compressing threads would crash with SIGSEGV.
What is the reason that we can't use a `ReentrantReadWriteLock` here to guard this race condition?
Performance concerns?

**Comment by luben** [comment]
No, it can't be increased. After `SHARED_LOCK_UPDATER.compareAndSet(this, 0, SHARED_LOCK_CLOSED)` the lock is closed and attempt to acquire it will throw exception. See https://github.com/luben/zstd-jni/blob/master/src/main/java/com/github/luben/zstd/AutoCloseBase.java#L25-L36

**Comment by luben** [comment]
Hmm, strange. I see some other methods that should be guarding the native code through the same lock but do not. May be that could be the issue, though these are not in the stack trace.

**Comment by luben** [comment]
added more guards around executing native code: https://github.com/luben/zstd-jni/commit/cec96538e692336eae11b033d9f09e7be30ff075 , but I am not sure if it will fix the problem here

**Comment by gaojieliu** [comment]
Thanks, we will give it a try when the new release is ready.

**Comment by luben** [comment]
just released 1.5.6-8 with that fix

**Comment by luben** [comment]
Closing as I assume it was fixed.
